File: platform/test/integration/providers/openai-embeddings/test_case.mjs
#!/usr/bin/env node

import path from 'node:path';
import { pathToFileURL } from 'node:url';

const args = process.argv.slice(2);
const readArg = (name, fallback) => {
  const idx = args.indexOf(name);
  return idx >= 0 ? args[idx + 1] : fallback;
};

const appEnv = readArg('--app-env', process.env.APP_ENV || 'dev');
const configDir = readArg('--config-dir', process.env.CONFIG_DIR);
if (!configDir) {
  console.error('Missing --config-dir');
  process.exit(2);
}

const platformRoot = path.resolve(configDir, '..', '..');
const importFile = async (absPath) => import(pathToFileURL(absPath).href);

const appYamlLoadMod = await importFile(path.join(platformRoot, 'polyglot/app_yaml_load/mjs/dist/index.js'));
const authConfigMod = await importFile(path.join(platformRoot, 'packages_mjs/auth_config/dist/index.js'));

const { config: appConfig } = await appYamlLoadMod.loadAppYamlConfig({ configDir, appEnv });
const providerConfig = appConfig.getNested(['providers', 'openai_embeddings'], {});
const authOptions = authConfigMod.buildSdkAuthOptions(providerConfig, 'bearer');
const apiKey = authConfigMod.resolveApiKey(providerConfig);

const configResult = {
  service: 'openai-embeddings',
  mode: 'direct-package-import',
  mjs_package: '@internal/auth-config',
  app_yaml_initialized: true,
  app_yaml_providers_openai_embeddings: providerConfig,
  resolved_config: {
    base_url: providerConfig.base_url ?? null,
    model: providerConfig.model ?? null,
    health_endpoint: providerConfig.health_endpoint ?? null,
    auth_type: authOptions?.type ?? null,
    has_api_key: apiKey != null,
  },
  uses_app_yaml_directly: true,
  note: 'auth_config resolves provider auth from AppYamlConfig + env/context.',
};

// Simple embedding test
let embeddingResult;
const effectiveApiKey = (apiKey && !apiKey.startsWith('{{')) ? apiKey : process.env.OPENAI_API_KEY;

if (!effectiveApiKey) {
  embeddingResult = { test: 'embedding', skipped: true, reason: 'no api key' };
} else {
  const baseUrl = providerConfig.base_url ?? 'https://api.openai.com/v1';
  const response = await fetch(`${baseUrl}/embeddings`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${effectiveApiKey}`,
    },
    body: JSON.stringify({
      model: 'text-embedding-3-small',
      input: 'Hello, world!',
    }),
  });

  const data = await response.json();
  const embedding = data.data?.[0]?.embedding ?? [];
  embeddingResult = {
    test: 'embedding',
    skipped: false,
    model: data.model,
    dimensions: embedding.length,
    usage_total_tokens: data.usage?.total_tokens,
    first_5_values: embedding.slice(0, 5),
    passed: embedding.length > 0,
  };
}

const connectionTest = embeddingResult.skipped
  ? { skipped: true, reason: embeddingResult.reason }
  : { skipped: false, connected: embeddingResult.passed, status: 'embedding_returned' };

const summary = { ...configResult, embedding_test: embeddingResult, connection_test: connectionTest };
console.log(JSON.stringify(summary, null, 2));

File: platform/test/integration/providers/openai-embeddings/test_case.py
#!/usr/bin/env python3
"""Direct package test for openai_embeddings provider auth config."""

from __future__ import annotations

import argparse
import json
import os

import re
import sys
from pathlib import Path

_HERE = Path(__file__).resolve().parent
_PLATFORM = _HERE
while _PLATFORM.name != "platform" and _PLATFORM != _PLATFORM.parent:
    _PLATFORM = _PLATFORM.parent

def _inject_paths():
    """Add polyglot/packages_py source dirs to sys.path from platform pyproject.toml."""
    toml = _PLATFORM / "pyproject.toml"
    if not toml.exists():
        return
    for line in toml.read_text().splitlines():
        m = re.search(r'path\s*=\s*"([^"]+)"', line)
        if not m:
            continue
        pkg_path = (_PLATFORM / m.group(1)).resolve()
        src = pkg_path / "src"
        sys.path.insert(0, str(src if src.is_dir() else pkg_path))

_inject_paths()

from app_yaml_load import load_app_yaml_config
from auth_config import build_sdk_auth_options, resolve_api_key

def _effective_api_key(provider_config):
    key = resolve_api_key(provider_config)
    if not key or key.startswith("{{"):
        key = os.environ.get("OPENAI_API_KEY")
    return key

def test_config_resolution(app_cfg) -> dict:
    provider_config = app_cfg.get_nested("providers", "openai_embeddings", default={})
    auth_options = build_sdk_auth_options(provider_config, default_auth_type="bearer")
    api_key = resolve_api_key(provider_config)

    return {
        "service": "openai-embeddings",
        "mode": "direct-package-import",
        "python_package": "auth_config",
        "app_yaml_initialized": True,
        "app_yaml_providers_openai_embeddings": provider_config,
        "resolved_config": {
            "base_url": provider_config.get("base_url"),
            "model": provider_config.get("model"),
            "health_endpoint": provider_config.get("health_endpoint"),
            "auth_type": auth_options.get("type") if auth_options else None,
            "has_api_key": api_key is not None,
        },
        "uses_app_yaml_directly": True,
        "note": "auth_config resolves provider auth from AppYamlConfig + env/context.",
    }

def test_embedding(app_cfg) -> dict:
    from openai import OpenAI

    provider_config = app_cfg.get_nested("providers", "openai_embeddings", default={})
    api_key = _effective_api_key(provider_config)
    base_url = provider_config.get("base_url", "https://api.openai.com/v1")

    if not api_key:
        return {"test": "embedding", "skipped": True, "reason": "no api key"}

    client = OpenAI(api_key=api_key, base_url=base_url)
    response = client.embeddings.create(
        model="text-embedding-3-small",
        input="Hello, world!",
    )

    embedding = response.data[0].embedding
    return {
        "test": "embedding",
        "skipped": False,
        "model": response.model,
        "dimensions": len(embedding),
        "usage_total_tokens": response.usage.total_tokens,
        "first_5_values": embedding[:5],
        "passed": len(embedding) > 0,
    }

def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-env", default="dev")
    parser.add_argument("--config-dir", required=True)
    args = parser.parse_args()

    app_cfg = load_app_yaml_config(config_dir=args.config_dir, app_env=args.app_env).config

    config_result = test_config_resolution(app_cfg)
    embedding_result = test_embedding(app_cfg)

    if embedding_result.get("skipped"):
        connection_test = {"skipped": True, "reason": embedding_result.get("reason")}
    else:
        connection_test = {"skipped": False, "connected": embedding_result.get("passed", False), "status": "embedding_returned"}

    summary = {**config_result, "embedding_test": embedding_result, "connection_test": connection_test}
    print(json.dumps(summary, indent=2))
    return 0

if __name__ == "__main__":
    raise SystemExit(main())
